File: src/main/java/gargoyle/netsquares/prop/simple/LongProperty.java
package gargoyle.netsquares.prop.simple;

import gargoyle.netsquares.prop.RWProperty;

import java.io.Serializable;
import java.util.Objects;

public final class LongProperty extends RWProperty<Long> implements Comparable<LongProperty>, Serializable {
    public LongProperty(Long value) {
        super(value);
    }

    public LongProperty() {
        this(0);
    }

    public LongProperty(long value) {
        super(value);
    }

    @Override
    public int compareTo(LongProperty o) {
        return Double.compare(get(), o.get());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RWProperty)) return false;
        if (!super.equals(o)) return false;
        RWProperty<?> that = (RWProperty<?>) o;
        return Objects.equals(get(), that.get());
    }

    public long getValue() {
        return get();
    }

    public void setValue(long value) {
        set(value);
    }
}

File: src/main/kotlin/gargoyle/netsquares/prop/simple/LongProperty.kt
package gargoyle.netsquares.prop.simple

import gargoyle.netsquares.prop.RWProperty

class LongProperty(value: Long = 0) : RWProperty<Long>(value), Comparable<LongProperty> {

    override fun compareTo(other: LongProperty): Int {
        return get().toDouble().compareTo(other.get().toDouble())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RWProperty<*>) return false
        if (!super.equals(other)) return false
        return get() == other.get()
    }

    var value: Long
        get() = get()
        set(value) {
            set(value)
        }

    override fun hashCode(): Int = get().hashCode()
}
